File: backend/internal/feature/prompt/domain/prompt.go
package domain

import (
	"errors"
	"time"
)

var (
	ErrNotFound  = errors.New("prompt not found")
	ErrForbidden = errors.New("prompt forbidden")
)

type Prompt struct {
	ID        int64
	UserID    *int64
	Name      string
	Body      string
	IsActive  bool
	IsDefault bool
	CreatedAt time.Time
	UpdatedAt time.Time
}

func (p *Prompt) OwnedBy(userID int64) bool {
	return p.UserID != nil && *p.UserID == userID
}

type CreateParams struct {
	UserID   int64
	Name     string
	Body     string
	IsActive bool
}

type PatchParams struct {
	Name     *string
	Body     *string
	IsActive *bool
}

type UpdateParams struct {
	ID       int64
	Name     string
	Body     string
	IsActive bool
}

File: ios/VoiceBlog/VoiceBlog/Models/Prompt.swift
import Foundation

struct Prompt: Identifiable, Sendable {
    let id: Int64
    let userId: Int64?
    let name: String
    let body: String
    let isActive: Bool
    let isDefault: Bool
    let createdAt: Date
    let updatedAt: Date

    var isSystemPrompt: Bool {
        userId == nil
    }

}

extension Prompt: Codable {
    nonisolated enum CodingKeys: String, CodingKey {
        case id
        case userId = "user_id"
        case name
        case body
        case isActive = "is_active"
        case isDefault = "is_default"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

struct PromptCreateRequest: Sendable {
    let name: String
    let body: String
    let isActive: Bool?
}

extension PromptCreateRequest: Codable {
    nonisolated enum CodingKeys: String, CodingKey {
        case name
        case body
        case isActive = "is_active"
    }
}

struct PromptUpdateRequest: Sendable {
    let name: String?
    let body: String?
    let isActive: Bool?
}

extension PromptUpdateRequest: Codable {
    nonisolated enum CodingKeys: String, CodingKey {
        case name
        case body
        case isActive = "is_active"
    }
}
